feat: 대댓글 관리 기능

jamie9504 opened the pull request:
**이슈 번호**

resolved: #242 

**작업 내용**

1. 대댓글 관리 기능 추가
2. 댓글 수정 기능 추가
3. 대댓글이 있는 댓글 삭제시, 댓글은 Deleted로 바뀌고, 필드들을 Null로 반환
4. Deleted인 상태의 댓글의 대댓글이 모두 삭제되는 경우, 해당 댓글도 삭제됨

**테스트 작성 여부**

- [X] Test case

**주의 사항**

jamie9504 committed:
feat: 댓글 수정기능 추가

jamie9504 committed:
Merge branch 'develop' of https://github.com/woowacourse-teams/2020-legeno-around-here into feature/#242

# Conflicts:
#	back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/service/CommentService.java

jamie9504 committed:
feat: Cocomment 등록 기능 추가 및 Comment Update 기능 관련 테스트 추가

jamie9504 committed:
test: 대댓글 생성 기능 서비스 테스트 추가

jamie9504 committed:
test: 대댓글 작성기능 인수테스트 추가

jamie9504 committed:
test: 대댓글, 댓글 순차 삭제의 경우 인수테스트 추가

jamie9504 committed:
feat: CommentState 추가

jamie9504 committed:
refactor: 댓글/대댓글 삭제 기능 개선

jamie9504 committed:
Merge branch 'develop' of https://github.com/woowacourse-teams/2020-legeno-around-here into feature/#242

jamie9504 committed:
Merge branch 'develop' of https://github.com/woowacourse-teams/2020-legeno-around-here into feature/#242

jamie9504 committed:
Merge branch 'develop' of https://github.com/woowacourse-teams/2020-legeno-around-here into feature/#242

jamie9504 committed:
feat: Flyway 반영

- Local & Test도 Flyway 가능하게끔 변경
- User Email 삭제되더라도 무조건 Unique하게 DB상 설정하여 예외 발생시 처리하는 로직 추가
- flyway DB Table 추가
- application.properties 정비
- SpringBootTest RandomPort 이용하도록 변경

aegis1920 commented on back-end/legeno-around-here/src/test/resources/application.properties:
```diff
@@ -1,9 +1,24 @@
-spring.flyway.enabled=false
+###################### DB SETTING ###############################
+spring.datasource.driver-class-name=org.h2.Driver
+spring.datasource.url=jdbc:h2:mem:testdb;MODE=Mysql;DB_CLOSE_DELAY=-1;DB_CLOSE_ON_EXIT=FALSE
```
url 뒤에 처음보는 옵션이 붙는데 있는 것과 없는 것의 차이점을 설명해주실 수 있을까요?

jamie9504 replied:
가장 중요한건 MODE=Mysql인데요, 이 부분이 있어야 Flyway가 지금 mysql 문법으로 작성되어있는데 이 DDL을 h2가 치환해서 받아들여줍니다.

그 외는 설정하지 않았을 때 기본으로 들어있는 설정들이라 붙여줬어요 [LINK](http://millky.com/@origoni/post/1155)

aegis1920 commented on back-end/legeno-around-here/src/main/resources/application-local.properties:
```diff
@@ -1,13 +1,29 @@
+spring.jpa.open-in-view=true
 spring.jpa.properties.hibernate.format_sql=true
+spring.jpa.properties.hibernate.show_sql=true
+spring.jpa.properties.hibernate.enable_lazy_load_no_trans=true
```
이 속성을 사용하신 이유가 있을까요?

jamie9504 replied:
Production 환경과 동일하게 맞춰보려고 했습니다!

`spring.jpa.properties.hibernate.show_sql`와 `spring.jpa.properties.hibernate.format_sql`의 경우 SQL console Log를 보기 위해 남겨놨습니다.

그런데 `enable_lazy_load_no_trans`는 사용하지 않는 것이 좋다하여 제거했습니다. Product도 머지할 때 제거해야할 것 같아요.

aegis1920 commented on back-end/legeno-around-here/src/main/resources/db/migration/V1__Initial_Setup.sql:
```diff
@@ -0,0 +1,135 @@
+CREATE TABLE `area`
```
작성하시느라 고생하셨습니다 👍 덕분에 flyway에 대한 부분도 테스트할 수 있게 됐네요! 
생각해봤는데 테스트 DB를 어차피 Drop할 거라면 `CREATE`하기 전에 `DROP TABLE IF EXISITS ~~`를 사용해도 될 것 같아요 :)

jamie9504 replied:
V1의 경우 맨 첫 작업이라서, 깨끗한 상황에서 실행된다는 것을 가정으로 하기 때문에 비워두는 것이 좋을 것 같아요!

있다면 예상하지 못한 경우라 에러가 나는 것이 좋다고 생각합니다.

aegis1920 commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/domain/comment/Comment.java:
```diff
@@ -91,28 +98,58 @@ private void validateAvailable() {
+        return !cocomments.isEmpty();
+    }
+
+    public boolean hasOnlyCocomment() {
+        return cocomments.size() == 1;
+    }
```
대댓글이 하나만 있는지 확인하는 로직이 필요한 이유가 있나요?

jamie9504 replied:
댓글 속성이 삭제이고 대댓글이 하나만 있을 때, 그 대댓글이 삭제되는 경우 삭제 속성의 댓글을 soft delete 하기 위함입니다.

aegis1920 reviewed (commented):
안녕하세요 제이미~
어쩌다보니 피드백보다는 궁금한 점이 더 많네요ㅎㅎ
고생하셨습니당 :)

jamie9504 committed:
Merge branch 'develop' of https://github.com/woowacourse-teams/2020-legeno-around-here into feature/#242

jamie9504 committed:
refactor: application.properties 속성 수정

- enable_lazy_load_no_trans 사용하지 않는 것이 좋다고 판단하여 제거(안티패턴)
- spring.datasource.driver-class-name, username, password 기본 값과 동일하여 제거

jamie9504 committed:
refactor: Flyway SQL명 이슈번호 추가 & 사용하지 않는 설정 제거

jamie9504 committed:
Merge branch 'develop' of https://github.com/woowacourse-teams/2020-legeno-around-here into feature/#242

Hamliet commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/domain/comment/Comment.java:
```diff
@@ -91,28 +98,58 @@ private void validateAvailable() {
+        return Objects.isNull(superComment);
+    }
+
+    public boolean hasCocoments() {
```
```suggestion
    public boolean hasCocomments() {
```
😎

jamie9504 replied:
👍 👍 👍 👍

Hamliet commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/service/CommentService.java:
```diff
@@ -36,32 +42,50 @@ public CommentResponse createComment(Long postId, CommentRequest commentRequest)
-        Comment comment = findCommentBy(commentId);
+        Comment comment = findAvailableCommentBy(commentId);
         validateIsCreator(comment);
+        if (comment.hasCocoments()) {
```
```suggestion
        if (comment.hasCocomments()) {
```
😎

jamie9504 replied:
👍 👍 👍 👍 👍

Hamliet commented on back-end/legeno-around-here/src/main/resources/db/migration/V2__Issue_242_Add_Cocoment.sql:
```diff
@@ -0,0 +1,2 @@
+ALTER TABLE `comment`
+    ADD `super_comment_id` BIGINT(20);
```
`V2__Issue_242_Add_Cocoment.sql` => `V2__Issue_242_Add_Cocomment.sql `
😎

jamie9504 replied:
👍 👍 👍 👍

Hamliet commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/domain/comment/Comment.java:
```diff
@@ -91,28 +98,58 @@ private void validateAvailable() {
+        return !cocomments.isEmpty();
+    }
+
+    public boolean hasOnlyCocomment() {
+        return cocomments.size() == 1;
+    }
```
대댓글이 여러개 있을 때, 그 대댓글이 모두 삭제되는 경우도 있을 것 같아요. size 말고 list의 모든 cocoment의 state를 확인하는 로직이 필요할 것 같습니다.

jamie9504 replied:
댓글은 단 건 삭제밖에 존재하지 않고,

대댓글의 경우엔 삭제시 soft delete가 되어 사이즈가 감소됩니다.
결국 최종 대댓글이 1개가 남아있고, 그 대댓글이 삭제될 경우에, 상위 댓글이 삭제 상태라면 삭제되는 것이죠!

Hamliet reviewed (commented):
수고하셨습니다 제이미! 코멘트 확인 부탁드려요~! :)

jamie9504 committed:
refactor: 오타 수정

Co-authored-by: Hamliet <hamliet@hanyang.ac.kr>

jamie9504 committed:
refactor: 오타 수정

Co-authored-by: Hamliet <hamliet@hanyang.ac.kr>

jamie9504 committed:
refactor: 파일명 변경

jamie9504 committed:
Merge branch 'feature/#242' of https://github.com/jamie9504/2020-legeno-around-here into feature/#242

yelimkim98 commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/controller/CommentController.java:
```diff
@@ -30,6 +31,25 @@
+        @RequestBody CommentRequest commentRequest) {
+        CommentResponse commentResponse = commentService.createCocomment(commentId, commentRequest);
+        return ResponseEntity
+            .created(URI.create("/comments/" + commentResponse.getId()))
```
201 헤더에 들어가고있는 URI 가
해당 대댓글을 조회하는 URI 맞나요?!
`capzzang.co.kr/{이 URI}`
이렇게 들어가게되는 것 맞나요??!

yelimkim98 commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/service/UserService.java:
```diff
@@ -46,18 +47,15 @@
-
         Area area = findAreaBy(userCreateRequest.getAreaId());
         User user = UserAssembler.assemble(userCreateRequest, area);
-        User createdUser = userRepository.save(user);
-
-        return createdUser.getId();
+        try {
+            User createdUser = userRepository.save(user);
+            return createdUser.getId();
+        } catch (DataIntegrityViolationException e) {
+            throw new WrongUserInputException(
+                "[" + userCreateRequest.getEmail() + "] 이메일은 이미 사용중입니다.");
+        }
```
👍

yelimkim98 reviewed (approved):
수고하셨습니다~~~~

Hamliet reviewed (approved):
수고하셨습니다 제이미! 이만 Approve할게요~!
